#1387: Add `TextDirectoryCorpus` that yields one doc per file recursi…

macks22 opened the pull request:
…vely read from directory. Resolves #1387.

macks22 commented:
@piskvorky I know we had discussed refactoring the TextCorpus variants in general; do you think it's best to merge this and then open up another issue for the more general architectural changes we were discussing, or should I add that sort of stuff to this PR? Thanks!

macks22 commented:
@menshikh-iv do you believe this requires any changes? Also, any thoughts on the question I posed to @piskvorky?

menshikh-iv commented on gensim/corpora/textcorpus.py:
```diff
@@ -49,35 +65,104 @@ class TextCorpus(interfaces.CorpusABC):
 
     """
-    def __init__(self, input=None):
+    def __init__(self, input=None, metadata=False, character_filters=None, tokenizer=None,
```
Please, add documentation for a new parameters in [google-docstring format](https://github.com/RaRe-Technologies/gensim/wiki/Developer-page#docstrings)

macks22 replied:
done

menshikh-iv commented on gensim/corpora/textcorpus.py:
```diff
@@ -49,35 +65,104 @@ class TextCorpus(interfaces.CorpusABC):
     def __iter__(self):
-        """
-        The function that defines a corpus.
+        """The function that defines a corpus.
```
Please use google-docstring format everywhere

macks22 replied:
done

menshikh-iv commented on gensim/corpora/textcorpus.py:
```diff
@@ -125,5 +211,121 @@ def __len__(self):
+    """
+
+    def __init__(self, input, metadata=False, min_depth=0, max_depth=None, pattern=None,
+                 exclude_pattern=None, **kwargs):
```
Add docstring with parameter description

macks22 replied:
done

menshikh-iv commented on gensim/corpora/textcorpus.py:
```diff
@@ -125,5 +211,121 @@ def __len__(self):
+
+class TextDirectoryCorpus(TextCorpus):
+    """Read documents recursively from a directory,
+    where each file is interpreted as a plain text document.
```
Maybe iterate by line (not by file), what do you think about it?

macks22 replied:
I think we could add an option for that, but the corpora this was designed for (20 newsgroups, Wiki-Movie from coherence papers) are one document per file.

macks22 replied:
added a new argument `lines_are_documents` along with a test that shows usage

menshikh-iv commented:
Can you add short example of usage of this feature with non-trivial tree (as "integration" test)
Example of tree:
```
.
├── a_folder/
│        ├── 0.txt
├── b_folder/
│       ├── 1.txt
│       ├── 2.txt
│       ├── c_folder/
│               ├── 3.txt
```

macks22 commented:
@menshikh-iv I've addressed your PR comments; thank you for the review!

Sorry about the big blob of commits since your review -- I had to rebase and resolve several conflicts with gensim upstream. The first commit, d41b2c4, contains my changes in response to your review.

macks22 commented:
@menshikh-iv @piskvorky Looks like I botched up this PR with the rebase I did to resolve conflicts recently. I've opened up PR #1459 to replace it.
